itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCleanerWithReplication.java

```diff
@@ -0,0 +1,198 @@
+import java.io.IOException;
+import java.nio.file.Files;
+
+import static org.junit.Assert.assertEquals;
+
+public class TestCleanerWithReplication extends CompactorTest {
+  private Path cmRootDirectory;
+  private static FileSystem fs;
+  private static MiniDFSCluster miniDFSCluster;
+
+  @Before
+  public void setup() throws Exception {
+    conf = new HiveConf();
+    TxnDbUtil.setConfValues(conf);
+    TxnDbUtil.cleanDb(conf);
+    conf.set("fs.defaultFS", fs.getUri().toString());
+    conf.setBoolVar(HiveConf.ConfVars.REPLCMENABLED, true);
+    ms = new HiveMetaStoreClient(conf);
+    txnHandler = TxnUtils.getTxnStore(conf);
+    cmRootDirectory = new Path(conf.get(HiveConf.ConfVars.REPLCMDIR.varname));
+    if (!fs.exists(cmRootDirectory)) {
+      fs.mkdirs(cmRootDirectory);
+    }
+    tmpdir = new File(Files.createTempDirectory("compactor_test_table_").toString());
+  }
+
+  @BeforeClass
+  public static void classLevelSetup() throws LoginException, IOException {
+    Configuration hadoopConf = new Configuration();
+    hadoopConf.set("dfs.client.use.datanode.hostname", "true");
+    hadoopConf.set("hadoop.proxyuser." + Utils.getUGI().getShortUserName() + ".hosts", "*");
+    miniDFSCluster =
+        new MiniDFSCluster.Builder(hadoopConf).numDataNodes(1).format(true).build();
+    fs = miniDFSCluster.getFileSystem();
+  }
+
+  @After
+  public void tearDown() throws IOException {
+    fs.delete(cmRootDirectory, true);
+    compactorTestCleanup();
+  }
+
+  @AfterClass
+  public static void tearDownClass() {
+    miniDFSCluster.shutdown();
+  }
+
+  @Test
+  public void cleanupAfterMajorTableCompaction() throws Exception {
+    Table t = newTable("default", "camtc", false);
+
+    addBaseFile(t, null, 20L, 20);
+    addDeltaFile(t, null, 21L, 22L, 2);
+    addDeltaFile(t, null, 23L, 24L, 2);
+    addBaseFile(t, null, 25L, 25);
+
+    burnThroughTransactions("default", "camtc", 25);
+
+    CompactionRequest rqst = new CompactionRequest("default", "camtc", CompactionType.MAJOR);
+    txnHandler.compact(rqst);
+    CompactionInfo ci = txnHandler.findNextToCompact("fred");
+    txnHandler.markCompacted(ci);
+    txnHandler.setRunAs(ci.id, System.getProperty("user.name"));
+
+    assertCleanerActions(6);
+  }
+
+  @Test
+  public void cleanupAfterMajorPartitionCompaction() throws Exception {
+    Table t = newTable("default", "campc", true);
+    Partition p = newPartition(t, "today");
+
+    addBaseFile(t, p, 20L, 20);
+    addDeltaFile(t, p, 21L, 22L, 2);
+    addDeltaFile(t, p, 23L, 24L, 2);
+    addBaseFile(t, p, 25L, 25);
+
+    burnThroughTransactions("default", "campc", 25);
+
+    CompactionRequest rqst = new CompactionRequest("default", "campc", CompactionType.MAJOR);
+    rqst.setPartitionname("ds=today");
+    txnHandler.compact(rqst);
+    CompactionInfo ci = txnHandler.findNextToCompact("fred");
+    txnHandler.markCompacted(ci);
+    txnHandler.setRunAs(ci.id, System.getProperty("user.name"));
+
+    assertCleanerActions(6);
```

Comment:
Why the number of files 6 in some test and 4 in some? The number of base/delta files added looks same.

Reply:
thats what existing passing tests create in cleaner and using the same here.